File: DataEngineering/08_Data_Governance_and_Quality/projects/python/04_anomaly_alerting/main.py
import os
import numpy as np
from sklearn.ensemble import IsolationForest
from slack_sdk import WebClient
from slack_sdk.errors import SlackApiError

def detect_anomalies(data):
    """
    Uses IsolationForest to detect anomalies in a dataset.
    Isolation Forest is standard for outlier detection in high-dimensional data.
    """
    # Reshape data for sklearn (requires 2D array)
    X = np.array(data).reshape(-1, 1)
    
    # Initialize and fit the model
    # contamination indicates the expected proportion of outliers (e.g., 5%)
    model = IsolationForest(contamination=0.05, random_state=42)
    model.fit(X)
    
    # Predict anomalies: -1 for anomaly, 1 for normal
    predictions = model.predict(X)
    
    anomalies = []
    for i, pred in enumerate(predictions):
        if pred == -1:
            anomalies.append((i, data[i]))
            
    return anomalies

def send_slack_alert(anomalies):
    """
    Sends an alert to a Slack channel using slack_sdk.
    """
    slack_token = os.environ.get("SLACK_BOT_TOKEN")
    if not slack_token:
        print("SLACK_BOT_TOKEN environment variable not set. Skipping real Slack alert.")
        # Fallback to local logging
        for idx, val in anomalies:
            print(f"[MOCK ALERT] Anomaly detected at index {idx} with value {val}.")
        return

    client = WebClient(token=slack_token)
    channel_id = "#data-engineering-alerts"
    
    blocks = [
        {
            "type": "header",
            "text": {
                "type": "plain_text",
                "text": "🚨 Data Anomaly Alert 🚨"
            }
        },
        {
            "type": "section",
            "text": {
                "type": "mrkdwn",
                "text": f"Detected *{len(anomalies)}* anomalies in the recent data pipeline run."
            }
        }
    ]
    
    for idx, val in anomalies:
        blocks.append({
            "type": "section",
            "text": {
                "type": "mrkdwn",
                "text": f"• Index: `{idx}`, Value: `{val}`"
            }
        })
        
    try:
        response = client.chat_postMessage(
            channel=channel_id,
            blocks=blocks,
            text="Data Anomaly Alert"
        )
        print(f"Alert successfully sent to {channel_id}.")
    except SlackApiError as e:
        print(f"Error sending message to Slack: {e.response['error']}")

def main():
    print("Fetching recent pipeline metrics...")
    # Simulate a stream of data (e.g., row counts, latency, or total sales)
    # Most values are around 100, but 500 is a clear anomaly.
    historical_data = [100, 105, 95, 110, 90, 102, 98, 101, 104, 96, 500, 99, 100]
    
    print("Running Isolation Forest Anomaly Detection...")
    anomalies = detect_anomalies(historical_data)
    
    if anomalies:
        print(f"Found {len(anomalies)} anomalies. Triggering alerts...")
        send_slack_alert(anomalies)
    else:
        print("No anomalies detected. Pipeline is healthy.")

if __name__ == "__main__":
    main()

File: DataEngineering/08_Data_Governance_and_Quality/projects/golang/04_anomaly_alerting/main.go
package main

import (
	"fmt"
	"log"
	"math"
	"os"

	"github.com/slack-go/slack"
)

// calculateStats computes the mean and standard deviation of a dataset
func calculateStats(data []float64) (mean float64, stddev float64) {
	if len(data) == 0 {
		return 0, 0
	}
	var sum float64
	for _, v := range data {
		sum += v
	}
	mean = sum / float64(len(data))

	var varianceSum float64
	for _, v := range data {
		varianceSum += math.Pow(v-mean, 2)
	}
	variance := varianceSum / float64(len(data))
	stddev = math.Sqrt(variance)
	return mean, stddev
}

// detectAnomalies uses Z-Score to identify anomalies.
// A Z-Score threshold of 3 is typical in statistical quality control.
func detectAnomalies(data []float64, threshold float64) []int {
	mean, stddev := calculateStats(data)
	var anomalies []int

	for i, val := range data {
		zScore := math.Abs((val - mean) / stddev)
		if zScore > threshold {
			anomalies = append(anomalies, i)
		}
	}
	return anomalies
}

// sendSlackAlert pushes a notification to a Slack channel
func sendSlackAlert(data []float64, anomalies []int) {
	token := os.Getenv("SLACK_BOT_TOKEN")
	if token == "" {
		fmt.Println("SLACK_BOT_TOKEN environment variable not set. Skipping real Slack alert.")
		for _, idx := range anomalies {
			fmt.Printf("[MOCK ALERT] Anomaly detected at index %d (value: %.2f)\n", idx, data[idx])
		}
		return
	}

	api := slack.New(token)
	channelID := "#data-engineering-alerts" // Typically injected via config

	var blocks []slack.Block

	headerText := slack.NewTextBlockObject("plain_text", "🚨 Data Anomaly Alert 🚨", false, false)
	headerBlock := slack.NewHeaderBlock(headerText)
	blocks = append(blocks, headerBlock)

	msgText := slack.NewTextBlockObject("mrkdwn", fmt.Sprintf("Detected *%d* anomalies in recent metrics.", len(anomalies)), false, false)
	sectionBlock := slack.NewSectionBlock(msgText, nil, nil)
	blocks = append(blocks, sectionBlock)

	for _, idx := range anomalies {
		details := slack.NewTextBlockObject("mrkdwn", fmt.Sprintf("• Index: `%d`, Value: `%.2f`", idx, data[idx]), false, false)
		blocks = append(blocks, slack.NewSectionBlock(details, nil, nil))
	}

	_, _, err := api.PostMessage(
		channelID,
		slack.MsgOptionBlocks(blocks...),
		slack.MsgOptionText("Data Anomaly Alert", false),
	)
	if err != nil {
		log.Printf("Error sending Slack message: %v\n", err)
	} else {
		fmt.Printf("Alert successfully sent to %s\n", channelID)
	}
}

func main() {
	fmt.Println("Fetching recent pipeline metrics...")
	// Simulated metrics (e.g., number of failed records per minute)
	metrics := []float64{10, 12, 11, 9, 13, 10, 11, 100, 12, 10, 9}

	fmt.Println("Running Z-Score Anomaly Detection...")
	// Use 2.5 standard deviations as the threshold
	anomalies := detectAnomalies(metrics, 2.5)

	if len(anomalies) > 0 {
		fmt.Printf("Found %d anomalies. Triggering alerts...\n", len(anomalies))
		sendSlackAlert(metrics, anomalies)
	} else {
		fmt.Println("No anomalies detected. Pipeline is healthy.")
	}
}
